examples/virtual-dom-wip/platform/Effect.roc

```diff
@@ -0,0 +1,82 @@
+hosted Effect
+    exposes [
+        Effect,
+        after,
+        always,
+        map,
+        NodeId,
+        EventHandlerId,
+        eventHandlerId,
+        createElement,
+        createTextNode,
+        updateTextNode,
+        appendChild,
+        removeNode,
+        setAttribute,
+        removeAttribute,
+        setProperty,
+        removeProperty,
+        setListener,
+        removeListener,
+        enableVdomAllocator,
+        disableVdomAllocator,
+    ]
+    imports []
+    generates Effect with [after, always, map]
+
+# TODO: private type
+NodeId : Nat
+
+EventHandlerId := Nat
+eventHandlerId = \id -> @EventHandlerId id
+
+# TODO: make these tag unions to avoid encoding/decoding standard names
+TagName : Str
+AttrType : Str
+EventType : Str
+
+## createElement tagName
+createElement : TagName -> Effect NodeId
+
+## createTextNode content
+createTextNode : Str -> Effect NodeId
+
+## updateTextNode content
+updateTextNode : NodeId, Str -> Effect {}
+
+## appendChild parentId childId
+appendChild : NodeId, NodeId -> Effect {}
+
+## removeNode id
+removeNode : NodeId -> Effect {}
+
+## setAttribute nodeId attrName value
+setAttribute : NodeId, AttrType, Str -> Effect {}
+
+## removeAttribute nodeId attrName
+removeAttribute : NodeId, AttrType -> Effect {}
+
+## setProperty nodeId propName json
+setProperty : NodeId, Str, List U8 -> Effect {}
+
+## removeProperty nodeId propName
+removeProperty : NodeId, Str -> Effect {}
+
+## setListener nodeId eventType handlerId
+setListener : NodeId, EventType, EventHandlerId -> Effect {}
+
+## removeListener nodeId eventType
+removeListener : NodeId, EventType -> Effect {}
+
+# Enable a special memory allocator for virtual DOM
+# This consists of two arenas, "even" and "odd", which alternately hold the "old" and "new" VDOM.
+# After we do a diff, the "old" virtual DOM can be dropped without checking refcounts.
```

Comment:
How does the platform know when the odd arena can be dropped? Is that just when Roc returns-to-host?

Reply:
It's not implemented yet, just comments so far. But it would swap every time we do a diff.
On every diff you compare the old virtual DOM to the new one. As soon as that's done, the old one is never used again and can be dropped.